pkg/collections.go

```diff
@@ -0,0 +1,264 @@
+	s := spinner.New(spinner.CharSets[34], 100*time.Millisecond)
+	s.Start()
+	for _, sourceColl := range collections {
+		sourceColl := sourceColl
+		g.Go(func() error {
+			if err := sem.Acquire(ctx, 1); err != nil {
+				return err
+			}
+			defer sem.Release(1)
+			var props driver.CollectionProperties
+			c.backoffCall(ctx, func() error {
+				sourceProps, err := sourceColl.Properties(ctx)
+				if err != nil {
+					c.Logger.Error().Err(err).Msg("Failed to get properties.")
+					return err
+				}
+				props = sourceProps
+				return nil
+			})
+			if props.IsSystem {
+				// skip system collections
+				return nil
+			}
+			var destinationColl driver.Collection
+			c.backoffCall(ctx, func() error {
+				dColl, err := c.ensureDestinationCollection(ctx, destinationDB, sourceColl, props)
+				if err != nil {
+					c.Logger.Error().Err(err).Msg("Failed to ensure destination collection.")
+					return err
+				}
+				destinationColl = dColl
+				return nil
+			})
+			bindVars := map[string]interface{}{
+				"@c": sourceColl.Name(),
+			}
+			var cursor driver.Cursor
+			c.backoffCall(ctx, func() error {
+				cr, err := sourceDB.Query(readCtx, "FOR d IN @@c RETURN d", bindVars)
+				if err != nil {
+					c.Logger.Error().Err(err).Str("collection", sourceColl.Name()).Msg("Failed to query source database for collection.")
+					return err
+				}
+				cursor = cr
+				return nil
+			})
+			batch := make([]interface{}, 0, c.BatchSize)
+			for {
+				var d interface{}
+				if err := backoff.Retry(func() error {
+					if _, err := cursor.ReadDocument(readCtx, &d); err != nil {
+						c.Logger.Error().Err(err).Str("collection", sourceColl.Name()).Msg("Read documents failed.")
+						return err
+					}
+					batch = append(batch, d)
+					return nil
+				}, backoff.WithContext(backoff.WithMaxRetries(backoff.NewExponentialBackOff(), uint64(backoffMaxTries)), ctx)); err != nil {
+					c.Logger.Error().Err(err).Str("collection", sourceColl.Name()).Msg("Backoff eventually failed.")
+					cursor.Close()
+					return err
+				}
+
+				if (!cursor.HasMore() && len(batch) > 0) || len(batch) >= c.BatchSize {
+					if err := backoff.Retry(func() error {
+						if _, _, err := destinationColl.CreateDocuments(restoreCtx, batch); err != nil {
+							c.Logger.Error().Err(err).Str("collection", sourceColl.Name()).Interface("document", d).Msg("Creating a document failed.")
+							return err
+						}
+						batch = make([]interface{}, 0, c.BatchSize)
+						return nil
+					}, backoff.WithContext(backoff.WithMaxRetries(backoff.NewExponentialBackOff(), uint64(backoffMaxTries)), ctx)); err != nil {
+						c.Logger.Error().Err(err).Str("collection", sourceColl.Name()).Msg("Backoff eventually failed.")
+						cursor.Close()
+						return err
+					}
+				}
+
+				if !cursor.HasMore() {
+					break
+				}
+			}
+			cursor.Close()
+
+			// Copy over all indexes for this collection.
+			if err := c.copyIndexes(restoreCtx, sourceColl, destinationColl); err != nil {
```

Comment:
Why have this after adding docs?

Reply:
I first thought that fields must exist for indexes. I guess they don't and can just be created as is.